core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantStageScheduler.java

```diff
@@ -0,0 +1,529 @@
+
+            taskLifecycleListener.taskCreated(stageExecution.getFragment().getId(), task);
+
+            task.addStateChangeListener(taskStatus -> updateTaskStatus(taskStatus, exchangeSinkInstanceHandle));
+            task.start();
+        }
+
+        if (taskFinishedFuture != null && !taskFinishedFuture.isDone()) {
+            blocked = taskFinishedFuture;
+        }
+    }
+
+    public synchronized boolean isFinished()
+    {
+        return failure == null &&
+                taskSource != null &&
+                taskSource.isFinished() &&
+                queuedPartitions.isEmpty() &&
+                finishedPartitions.containsAll(partitionToTaskDescriptorMap.keySet());
+    }
+
+    public void cancel()
+    {
+        close(false);
+    }
+
+    public void abort()
+    {
+        close(true);
+    }
+
+    private void fail(Throwable t)
+    {
+        synchronized (this) {
+            if (failure == null) {
+                failure = t;
+            }
+        }
+        close(true);
+    }
+
+    private void close(boolean abort)
+    {
+        boolean closed;
+        synchronized (this) {
+            closed = this.closed;
+            this.closed = true;
+        }
+        if (!closed) {
+            cancelRunningTasks(abort);
+            cancelBlockedFuture();
+            releaseAcquiredNode();
+            closeTaskSource();
+            closeSinkExchange();
+        }
+    }
+
+    private void cancelRunningTasks(boolean abort)
+    {
+        List<RemoteTask> tasks;
+        synchronized (this) {
+            tasks = ImmutableList.copyOf(runningTasks.values());
+        }
+        if (abort) {
+            tasks.forEach(RemoteTask::abort);
+        }
+        else {
+            tasks.forEach(RemoteTask::cancel);
+        }
+    }
+
+    private void cancelBlockedFuture()
+    {
+        verify(!Thread.holdsLock(this));
+        ListenableFuture<Void> future;
+        synchronized (this) {
+            future = blocked;
+        }
+        if (future != null && !future.isDone()) {
+            future.cancel(true);
+        }
+    }
+
+    private void releaseAcquiredNode()
+    {
+        verify(!Thread.holdsLock(this));
+        ListenableFuture<InternalNode> future;
+        synchronized (this) {
+            future = acquireNodeFuture;
+            acquireNodeFuture = null;
+        }
+        if (future != null) {
+            future.cancel(true);
+            if (future.isDone() && !future.isCancelled()) {
+                nodeAllocator.release(getFutureValue(future));
+            }
+        }
+    }
+
+    private void closeTaskSource()
+    {
+        TaskSource taskSource;
+        synchronized (this) {
+            taskSource = this.taskSource;
+        }
+        if (taskSource != null) {
+            try {
+                taskSource.close();
+            }
+            catch (RuntimeException e) {
+                log.warn(e, "Error closing task source for stage: %s", stageExecution.getStageId());
```

Comment:
Why not rethrow?

Reply:
Not sure if it makes sense to propagate this failure. It probably is not going to be useful and it would make reasoning about close correctness more difficult.